L73-binderchannel/L73-binderchannel.md

```diff
@@ -0,0 +1,317 @@
+# Draft of L73-binder-channel.md
+
+## BinderChannel
+
+*   Author(s): markb74
+*   Approver:
+*   Status: Draft
+*   Implemented in: Java
+*   Last updated: July 10, 2020
+*   Discussion at:
+
+## Abstract
+
+Proposes a gRPC channel and server for cross-process communication on Android,
+with an underlying transport using native Android [Binder] transactions.
+
+## Background
+
+While the majority of Android Apps have no need for cross-process communication,
+those that do are often large and complex, with dozens or even hundreds of
+contributors.
+
+The Android [bound service] API is relatively low-level, and leaves problems
+like threading, flow-control, and error-handling to the application developer.
+
+Just as the use of gRPC for network communication significantly reduces the
+burden on application developers vs. lower-level APIs (e.g. posix sockets), we
+expect the same to hold for cross-process communication on Android.
+
+Protocol buffers are also more resilient to version skew than traditional
+Android [Parcelable] instances, an important consideration when calls are
+cross-application.
+
+Despite this, the use of [bound services] is necessary for the Android platform
+to be aware of inter-process dependencies, and there's occasionally a need to
+pass platform-define parcelable objects (e.g. `PendingIntent`), some of which
+can't simply be serialized into a byte stream.
+
+## Proposal
+
+We propose the creation of client and server transport that supports
+communication via [Binder] transactions with an Android [bound service], with
+corresponding channel and server builders. This transport will support the
+inclusion of Android Parcelable objects in Metadata, to allow Parcelables when
+necessary, while discouraging Parcelables as a general message format.
```

Comment:
Not using Parcelable for serialization/deserialization is understandable because of versioning issues described above, but I am not sure there is enough justification for why Parcelables should be Metadata. Metadata is sent as headers/trailers and so has different semantics. Can one convert Parcelable to protobuf (and reverse) and include as part of request/response or is that also discouraged?

Reply:
It's because "some of which can't simply be serialized into a byte stream." It is not possible to form a protobuf encoding; they are handled specially by the OS. They have features similar to fd-passing for unix-domain sockets.

Reply:
Okay the UDS FD example and special handling by OS make sense. Otherwise I was thinking Metadata is "metadata" about an RPC (such as tracing/telemetry or routing info) but not the content of the RPC itself. So I got the impression that the limitations of serializing/encoding is why we "promoted" Parcelable to be metadata.

Reply:
FWIW, the transport _does_ actually support Parcelable messages internally (it's also documented in the wire format), so someone could create a marshaller to send Parcelable messages if they really wanted.

That gets quite ugly though, and you don't get to use generated stubs, so we want to discourage it.